File: aoj/src/ALDS1_6_C/main.py
from copy import copy

class Card:
    def __init__(self, input):
        suit, number = input.split(" ")
        self.suit = suit
        self.number = int(number)
    
    def __str__(self):
        return str(self.suit) + " " + str(self.number)

def swap(a, i, j):
    a[i], a[j] = a[j], a[i]

def partition(a, p, r):
    x = a[r].number
    i = p
    for j in range(p, r):
        if a[j].number <= x:
            swap(a, j, i)
            i += 1
    swap(a, i, r)
    return i

def quickSort(a, p, r):
    if p >= r:
        return
    i = partition(a, p, r)
    quickSort(a, p, i - 1)
    quickSort(a, i + 1, r)

def merge(a, left, mid, right):
    l = copy(a[left:mid])
    r = copy(a[mid:right])
    l.append(Card("c 1000000000"))
    r.append(Card("c 1000000000"))

    li, ri = 0, 0
    for i in range(left, right):
        if l[li].number <= r[ri].number:
            a[i] = l[li]
            li += 1
        else:
            a[i] = r[ri]
            ri += 1

def merge_sort(a, left, right):
    if left+1 >= right:
        return
    mid = int((left + right) / 2)
    merge_sort(a, left, mid)
    merge_sort(a, mid, right)
    merge(a, left, mid, right)
    return

def main():
    n = int(input())
    a = []
    b = []
    for _ in range(n):
        card = Card(input())
        a.append(card)
        b.append(card)
    quickSort(a, 0, n-1)
    merge_sort(b, 0, n)
    aa = [ str(v) for v in a ]
    ba = [ str(v) for v in b ]
    if aa == ba:
        print("Stable")
    else:
        print("Not stable")
    for c in a:
        print(c)


if __name__ == '__main__':
    main()

File: aoj/src/ALDS1_6_C/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"reflect"
	"strconv"
	"strings"
)

const max = 1000000000

type card struct {
	rawText string
	number  int
}

func newCard(s string) *card {
	splited := strings.Split(s, " ")
	number, err := strconv.Atoi(splited[1])
	if err != nil {
		panic(err)
	}
	return &card{
		rawText: s,
		number:  number,
	}
}

// ALDS1_6_C:クイックソート
func main() {
	stdin := bufio.NewScanner(os.Stdin)
	stdin.Buffer([]byte{}, math.MaxInt64)
	n := scanInt(stdin)
	cards := make([]*card, n, n)
	for i := 0; i < n; i++ {
		stdin.Scan()
		t := stdin.Text()
		cards[i] = newCard(t)
	}
	cards2 := make([]*card, n)
	copy(cards2, cards)
	quickSort(cards, 0, n-1)
	mergesort(cards2, 0, n)
	if reflect.DeepEqual(cards, cards2) {
		fmt.Println("Stable")
	} else {
		fmt.Println("Not stable")
	}
	for _, c := range cards {
		fmt.Println(c.rawText)
	}
}

func quickSort(a []*card, p, r int) {
	if p < r {
		q := partition(a, p, r)
		quickSort(a, p, q-1)
		quickSort(a, q+1, r)
	}

}

func partition(a []*card, p, r int) int {
	x := a[r].number
	i := p
	for j := p; j < r; j++ {
		if a[j].number <= x {
			swap(a, i, j)
			i++

		}
	}
	swap(a, i, r)
	return i

}

func swap(a []*card, i, j int) {
	tmp := a[i]
	a[i] = a[j]
	a[j] = tmp

}

func scanInt(sc *bufio.Scanner) int {
	sc.Scan()
	t := sc.Text()
	i, err := strconv.Atoi(t)
	if err != nil {
		panic(err)
	}
	return i
}

func scanIntArray(sc *bufio.Scanner) []int {
	sc.Scan()
	t := sc.Text()
	ss := strings.Split(t, " ")
	is := []int{}
	for _, s := range ss {
		i, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		is = append(is, i)
	}
	return is
}

func mergesort(a []*card, left, right int) {
	if left+1 >= right {
		return
	}
	mid := (left + right) / 2
	mergesort(a, left, mid)
	mergesort(a, mid, right)
	merge(a, left, mid, right)
	return
}

func merge(a []*card, left, mid, right int) {
	ls := a[left:mid]
	l := make([]*card, len(ls))
	copy(l, ls)
	rs := a[mid:right]
	r := make([]*card, len(rs))
	copy(r, rs)

	l = append(l, &card{number: max + 1})
	r = append(r, &card{number: max + 1})

	li, ri := 0, 0
	for i := left; i < right; i++ {
		if l[li].number <= r[ri].number {
			a[i] = l[li]
			li++
		} else {
			a[i] = r[ri]
			ri++
		}
	}

}
